Implement `#[deprecated]` attribute (RFC 1270)

petrochenkov opened the pull request:
Closes https://github.com/rust-lang/rust/issues/29935

The attributes `deprecated` and `rustc_deprecated` are completely independent in this implementation and it leads to some noticeable code duplication. Representing `deprecated` as
```
Stability {
    level: Stable { since: "" },
    feature: "",
    depr: Some(Deprecation),
}
```
or, contrariwise, splitting rustc_deprecation from stability makes most of the duplication go away.
I can do this refactoring, but before doing it I must be sure, that further divergence of `deprecated` and `rustc_deprecated` is certainly not a goal.

cc @llogiq

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

llogiq commented:
Great work, @petrochenkov! I think *allowing* the features to diverge further was the express intent of the RFC. This means we can use `#[rustc_deprecation]` as is and iterate on the `#[deprecation]` feature, e.g. adding other optional fields.

petrochenkov commented:
I'd prefer the standard library to migrate to `#[deprecated]` eventually (after its stabilization, for example), maybe with slightly stricter checks, like now. I.e. I see `#[rustc_deprecated]` as a temporary measure.

brson commented:
This patch looks quite thorough. Good work.

brson commented on src/librustc/middle/stability.rs:
```diff
@@ -165,9 +171,35 @@ impl<'a, 'tcx: 'a> Annotator<'a, 'tcx> {
+                                               "`#[deprecated]` attribute is unstable");
+                        fileline_help!(self.tcx.sess, attr.span(), "add #![feature(deprecated)] to \
+                                                                    the crate features to enable");
+                    }
```
This check should probably be in libsyntax/feature_gate.rs

brson commented on src/librustc/middle/stability.rs:
```diff
@@ -663,7 +711,40 @@ fn lookup_uncached<'tcx>(tcx: &ty::ctxt<'tcx>, id: DefId) -> Option<&'tcx Stabil
+                // unmarked impls for it. See FIXME above for more details.
+
+                debug!("lookup: trait_id={:?}", trait_id);
+                return lookup_deprecation(tcx, trait_id);
```
The duplication of this method strikes me as particularly bad. Can the contents be abstracted into `lookup_uncached_generic`?

petrochenkov replied:
I've removed some obsolete code from these functions instead, should be better now.

brson commented on src/librustc_lint/builtin.rs:
```diff
@@ -577,74 +577,71 @@ impl LateLintPass for MissingDebugImplementations {
     DEPRECATED,
     Warn,
-    "detects use of #[rustc_deprecated] items"
+    "detects use of `#[deprecated]` or `#[rustc_deprecated]` items"
```
The existence of `#[rustc_deprecated]` is an implementation detail. Can we change this string to 'detects use of deprecated items'?

brson commented on src/test/compile-fail/deprecation-in-staged-api.rs:
```diff
@@ -0,0 +1,18 @@
+
+// #[deprecated] can't be used in staged api
+
+#![feature(deprecated, staged_api)]
```
Didn't we remove the staged_api feature?

petrochenkov replied:
We removed `#[staged_api]` attribute, not `#[feature(staged_api)]`.
`#[feature(staged_api)]` gates stability attributes.

brson commented on src/test/auxiliary/deprecation-lint.rs:
```diff
@@ -0,0 +1,87 @@
+    fn trait_deprecated(&self) {}
+    #[deprecated(since = "1.0.0", note = "text")]
+    fn trait_deprecated_text(&self) {}
+}
```
Why isn't there a test with the deprecated attribute on the trait itself?

petrochenkov commented:
Updated with comments addressed.
I've also added rustdoc support for `#[deprecated]`. (I intended to avoid doing this initially and merge `Deprecation` into `Stability` instead.)

brson commented:
@bors r+

bors commented:
:pushpin: Commit 67a9784 has been approved by `brson`

<!-- @bors r=brson 67a978411a2009093bb40c4f1320a08d2a28b6c7 -->

bors commented:
:hourglass: Testing commit 67a9784 with merge ac2c5ff...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7432), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7450), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/706), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7388), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7408), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7392), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/618), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2571), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7462), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7453), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7419), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2419), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2425), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2428), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2416), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1672), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2409)

durka commented:
Is there a reason (heh) that this was implemented with `#[deprecated(note = "...")]` instead of `#[deprecated(reason = "...")]` as in [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/1270-deprecation.md#detailed-design)? Should we amend the RFC to match (the rename was mentioned as an alternative but not in the main text)?

petrochenkov commented:
@durka 
>Is there a reason (heh) that this was implemented with #[deprecated(note = "...")] instead of #[deprecated(reason = "...")] as in the RFC?

Yes, by request from the RFC author, also see https://github.com/rust-lang/rfcs/pull/1270#issuecomment-158149633
RFC should really be amended since it's a central piece of documentation about this feature at the moment.

liigo commented:
I'm surprised. Implementation **violate** the RFC (`note` VS `reason`) **intentionally** without any reason, and surprised that was merged without any query. I don't oppose `note` (I like it), I just obey the RFC. 

(This is not the first time i noticed that RFC was violating by intention. Please respect RFC! Please amend RFC **first** if you think there is an error.)

llogiq commented:
I'm a bit busy ATM but will push a PR to change the RFC shortly to alleviate the situation.

llogiq commented:
See [RFC PR #1425](https://github.com/rust-lang/rfcs/pull/1425)

durka commented:
Oops, I had already submitted the same PR at rust-lang/rfcs#1423. @liigo your language is a bit stronger than necessary since this decision was essentially made in the RFC thread as it was being accepted, but just didn't make it into the text on time.

liigo commented:
> since this decision was essentially made in the RFC thread

I don't think so. Only three people says "`note` seems fine", no consensus, no confirmed by the libs (or lang) team.

llogiq commented:
It isn't unusual to tweak details in the implementation after an RFC is accepted. Please forgive my cavalier attitude, but I don't think changing a field name on an unstable attribute requires the whole teams' consenus.

petrochenkov commented:
I've just noticed, this is not marked with "relnotes", but probably should be.

bluss commented:
Is there a tracking issue for stabilization?
